### PR TITLE
Change the fetch procedure of the installer

### PR DESCRIPTION
**Pull request by dmitryilyin**
The installer uses checkout to switch to the specified branch.
This method usually works, but is not compatible with Gerrit's
refs/changes/\* branches. Using "git fetch origin $ref" and
then "git checkout FETCH_HEAD" works with both normal and
Gerrit branches and is already being used for the "update"
action.

### CODE REVIEW

**Comment by tobias-urdin** [comment]
+1 on this, we have been forking librarian-puppet-simple for a long time to get this.

**Comment by jeffmccune** [comment]
I'm OK with merging this now, but I'd like to better understand the use case you have @dmitryilyin and @tobias-urdin .  The update subcommand results in a detached head state which makes it difficult to work on component modules within the development workflow.  I'm looking for ways to improve this.

It seems to me if the ref is a branch, then the branch should be checked out, not a detached head.  Your thoughts?

File: lib/librarian/puppet/simple/installer.rb [hunk 1]
```diff
@@ -53,7 +53,10 @@ def install_git(module_path, module_name, repo, ref = nil)
             system_cmd("git clone #{repo} #{module_name}")
             Dir.chdir(module_dir) do
               system_cmd('git branch -r')
-              system_cmd("git checkout #{ref}") if ref
+              if ref
+                system_cmd("git fetch origin #{ref}")
```
Comment (jeffmccune) [inline lib/librarian/puppet/simple/installer.rb]: I also _really_ don't like how we're hard coding a remote named `origin` here, but this is something that needs to be improved in a separate patch.

**Comment by tobias-urdin** [comment]
Thanks for looking into it!
We use Gerrit[1] for our code review which we integrate with Jenkins for our CI, changes pushed to Gerrit for code review is not merged but exists in a separate ref.

Git checkout does not support a ref but only branches. This change gives the possibility to use the `:ref` key in the `Puppetfile` to reference any ref available in git. That could be tags `ref/tags/mytag`, a simple branch or a custom ref which for us would be Jenkins pulling the `ref/changes/<change number>` ref that Gerrit has staged the change in but not merged into the branch yet.

To answer your question, my git foo is not expert level so I don't really know how we would grab a refspec without getting it detached. I can't find that it supports multiple origins but I agree hardcoding the origin might not be the best solution.

Best regards

[1] https://www.gerritcodereview.com/

**Comment by tobias-urdin** [comment]
@jeffmccune Is it possible to get a new release on Ruby Gems which this change? Seems like @bodepd is the owner so perhaps only he can push a new version.

**Comment by tobias-urdin** [comment]
Sorry, don't want to be a nuisance but a ping on the above question for releasing a new gem.
